Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,220 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+  ( SBVector(..)
+  , DBVector(..)
+  , SBMatrix(..)
+  , dot
+  , mult
+  , size
+  , linearSolve
+  , testLinearSolver
+  , getRandomSBVectors
+  ) where
+
+import qualified Data.List as L
+import qualified Data.Vector as V
+import qualified Data.Vector.Unboxed as U
+import qualified Data.Vector.Unboxed.Sized as SU
+import qualified Data.Vector.Unboxed.Mutable as MU
+import Control.Monad.ST
+import Debug.Trace
+import Data.Semigroup()
+import System.Random
+import System.IO.Unsafe
+import GHC.Clock
+import Data.Bit
+import Data.Bits
+import Data.Foldable
+import GHC.TypeNats
+import Data.Maybe
+import Data.Proxy
+
+-- Sparse Binary Vector
+newtype SBVector = SBVector
+  { unSBVector :: U.Vector Int
+  } deriving (Show)
+
+-- Dense Binary Vector
+-- newtype DBVector = DBVector
+--   { unDBVector :: U.Vector Bit
+--   } deriving (Show, Eq)
+
+newtype DBVector (k :: Nat) = DBVector
+  { unDBVector :: SU.Vector k Bit
```

Comment:
We can actually push this approach further (and it is a good exercise), encoding in types that the matrix is a square one. 

```haskell
import GHC.TypeNats hiding (Mod)
import Data.Mod.Word
import Unsafe.Coerce

-- Sparse Binary Vector
newtype SBVector (k :: Nat) = SBVector { unSBVector :: U.Vector (Mod k) }
  deriving (Show)

-- Dense Binary Vector
newtype DBVector (k :: Nat) = DBVector { unDBVector :: SU.Vector k Bit }
  deriving (Eq, Show)

-- Sparse Binary Matrix
newtype SBMatrix (k :: Nat) = SBMatrix { unSBMatrix :: SV.Vector k (SBVector k) }
  deriving (Show)

listBits' :: KnownNat k => DBVector k -> [Mod k]
listBits' = unsafeCoerce listBits

flipBit' :: KnownNat k => MUS.MVector k s Bit -> Mod k -> ST s ()
flipBit' = unsafeCoerce (unsafeFlipBit :: MU.MVector s Bit -> Int -> ST s ())

index' :: KnownNat k => SBMatrix k -> Mod k -> SBVector k
index' = unsafeCoerce SV.index
```

(There is nothing `unsafe` in these coercions, I'm just lazy to jump through all `newtype` hoops)

Put these extra deps into `stack.yaml`:

```yaml
extra-deps:
- random-1.2.0
- smallcheck-1.2.0
- splitmix-0.1
- github: Bodigrim/mod
  commit: 64e3b10
```